### DISCUSSION TITLE
KeyboardEvent fired from Stage for iOS

### DISCUSSION THREAD

**Discussion by toysonlocation** [airsdk/Adobe-Runtime-Support#1321] [Q&A] [answered]
I've been trying to get keypress events from the Stage on iOS for a long time. I have hacked together StageText, focus, and timer suppression of the softkeyboard to sort of get what I need (game-like access to a BT keyboard controller), but it is heavy and brittle. I just ported my application to Android and, thanks be to someone, found that stage.addEventListener(KeyboardEvent.KEY_DOWN) works! So, it would be huge to have this work on iOS, and I know I'm not the only one.

Is this planned, or can someone explain why it can't be done?

**Comment by ajwfrost** [comment]
Curiously, I got the same request from another customer just a couple of days ago, and I have to confess that I thought this was already possible. We'll look into it - not sure if there's a technical reason why it can't be done, I would have thought it should work!

**Reply by toysonlocation** [reply]
Thanks Andrew for having a look, it really would be a very significant improvement if it can be done.

**Reply by toysonlocation** [reply]
My new affair with Android and it's lovely handling of keypresses has come to an abrupt end. I've been cast back onto iOS cuz client reasons, and I must reintroduce all my icky hacks to get BT keypresses working again. Sorry for the fast re-ask: any luck seeing why iOS AIR doesn't provide KeyboardEvents from the Stage without a focused TextField?

**Reply by ajwfrost** [reply]
We're investigating this at the moment. Apple certainly don't make it easy: it seems like the way people tend to do this is just a hack very similar to what you could currently do with AS3 and text fields. We are trying to see whether we can have a hidden text field ourselves that's only then activated when a bluetooth keyboard is connected: we wouldn't be able to receive "key down" and "key up" events, so it would just be a case of "someone typed a character"...

**Reply by toysonlocation** [reply]
Reading this, https://developer.apple.com/documentation/uikit/mac_catalyst/handling_key_presses_made_on_a_physical_keyboard, I presume that AIR doesn't participate in a responder chain that could tap this event? Maybe AIR doesn't use UIKit, but there's an equivalent in the framework it does use? My Apple dev skills are minimal.

**Reply by ajwfrost** [reply]
Yes we'd been trying that one: we're getting the touch events, like you would expect, when someone pressed the screen: but we're not getting keyboard events. The closest we've found is to set up a responder that also has the UIKeyInput protocol but the problem with that is that if a hardware keyboard isn't attached, it will pop up the virtual keyboard.

We'll double-check this one though, and we're also investigating whether we could just set up "keyboard shortcuts" for a set of all keys (but that may cause massive side-effects...)

thanks

**Reply by toysonlocation** [reply]
If I could couple `SoftKeyboardEvent.SOFT_KEYBOARD_ACTIVATING => preventDefault()` with your UIKeyInput method, perhaps I would be happy? Ideally that would come with the ability to check if there are any "keyboards" attached to the device.

**Reply by ajwfrost** [reply]
I'm not 100% sure of the possibilities around that (cancelling the keyboard pop-up etc) but it does now look like we may be able to capture the raw key input, it seems that this (the original idea) works on more recent iOS versions. We'll have to try building this into the SDK and then giving it a go....

**Comment by ajwfrost** [comment]
FYI it looks like this capability is possible but only relatively recently: iOS 13.4 introduced a key object that means we can capture information from a full keyboard (prior to this, it was only allowing a few codes through for Apple TV remote controls by the looks of things).

So we've just added the appropriate code to capture events from a physical keyboard and to pass these to the runtime; assuming you don't have an input text area active then they would be dispatched from the root i.e. the stage object.

**Reply by toysonlocation** [reply]
That sounds fantastic! When might I be able to try this out?

**Reply by ajwfrost** [reply]
Target is the end of the day tomorrow...
thanks

**Reply by toysonlocation** [reply]
You guys rock

**Comment by toysonlocation** [comment, marked as answer]
Works in just released 33.1.1.686. It's so wonderful having this functionality after years of agony. Another gap this new feature fills is that if a TextField is focused, the Stage still fires the KeyboardEvent and even includes arrow keys. At first blush, it's basically everything I would have needed before. Thanks Andrew and team!
